File: packages/crypto/src/nip17.rs
//! NIP-17 Gift Wrap Implementation
//!
//! Provides metadata protection for private messages through:
//! 1. Rumor (kind 14): Unsigned inner message
//! 2. Seal (kind 13): Encrypted rumor, signed by sender
//! 3. Gift Wrap (kind 1059): Encrypted seal, signed by ephemeral key

use crate::error::CryptoError;
use crate::keys::{generate_keypair, get_public_key};
use crate::nip44::{nip44_decrypt, nip44_encrypt};
use crate::nostr::{compute_event_id, sign_event, verify_event, NostrEvent, UnsignedEvent};
use crate::randomize_timestamp;

/// Result of unwrapping a gift-wrapped message
#[derive(Debug, Clone)]
pub struct UnwrapResult {
    pub rumor: NostrEvent,
    pub sender_pubkey: String,
    pub seal_verified: bool,
}

/// Event kinds for NIP-17
const KIND_SEAL: i32 = 13;
const KIND_RUMOR: i32 = 14;
const KIND_GIFT_WRAP: i32 = 1059;

/// Time randomization range (2 days in seconds)
const TIMESTAMP_RANGE: u32 = 172800;

/// Create a rumor event (unsigned kind 14)
pub fn create_rumor(
    sender_pubkey: String,
    recipient_pubkey: String,
    content: String,
    created_at: i64,
) -> Result<NostrEvent, CryptoError> {
    // Randomize timestamp within ±2 days
    let randomized_time = randomize_timestamp(created_at, TIMESTAMP_RANGE);

    // Create unsigned event
    let unsigned = UnsignedEvent {
        pubkey: sender_pubkey,
        created_at: randomized_time,
        kind: KIND_RUMOR,
        tags: vec![vec!["p".to_string(), recipient_pubkey]],
        content,
    };

    // Compute ID but leave signature empty
    let id = compute_event_id(unsigned.clone())?;

    Ok(NostrEvent {
        id,
        pubkey: unsigned.pubkey,
        created_at: unsigned.created_at,
        kind: unsigned.kind,
        tags: unsigned.tags,
        content: unsigned.content,
        sig: String::new(), // Rumor has no signature
    })
}

/// Create a seal event (kind 13) containing encrypted rumor
pub fn create_seal(
    sender_private_key: Vec<u8>,
    recipient_pubkey: String,
    rumor: NostrEvent,
    created_at: i64,
) -> Result<NostrEvent, CryptoError> {
    // Serialize rumor to JSON
    let rumor_json = serialize_event(&rumor)?;

    // Encrypt rumor with NIP-44
    let encrypted_rumor = nip44_encrypt(
        sender_private_key.clone(),
        recipient_pubkey.clone(),
        rumor_json,
    )?;

    // Randomize timestamp
    let randomized_time = randomize_timestamp(created_at, TIMESTAMP_RANGE);

    // Get sender pubkey
    let sender_pubkey = get_public_key(sender_private_key.clone())?;

    // Create and sign seal
    let unsigned = UnsignedEvent {
        pubkey: sender_pubkey,
        created_at: randomized_time,
        kind: KIND_SEAL,
        tags: vec![], // No tags on seal
        content: encrypted_rumor,
    };

    sign_event(sender_private_key, unsigned)
}

/// Create a gift wrap event (kind 1059) containing encrypted seal
pub fn create_gift_wrap(
    recipient_pubkey: String,
    seal: NostrEvent,
    created_at: i64,
) -> Result<NostrEvent, CryptoError> {
    // Generate ephemeral keypair
    let ephemeral = generate_keypair();

    // Serialize seal to JSON
    let seal_json = serialize_event(&seal)?;

    // Encrypt seal with ephemeral key
    let encrypted_seal = nip44_encrypt(
        ephemeral.private_key.clone(),
        recipient_pubkey.clone(),
        seal_json,
    )?;

    // Randomize timestamp
    let randomized_time = randomize_timestamp(created_at, TIMESTAMP_RANGE);

    // Create and sign gift wrap with ephemeral key
    let unsigned = UnsignedEvent {
        pubkey: ephemeral.public_key,
        created_at: randomized_time,
        kind: KIND_GIFT_WRAP,
        tags: vec![vec!["p".to_string(), recipient_pubkey]],
        content: encrypted_seal,
    };

    sign_event(ephemeral.private_key, unsigned)
}

/// Unwrap a gift-wrapped message
pub fn unwrap_gift_wrap(
    recipient_private_key: Vec<u8>,
    gift_wrap: NostrEvent,
) -> Result<UnwrapResult, CryptoError> {
    // Verify gift wrap kind
    if gift_wrap.kind != KIND_GIFT_WRAP {
        return Err(CryptoError::InvalidCiphertext);
    }

    // Decrypt gift wrap to get seal
    let seal_json = nip44_decrypt(
        recipient_private_key.clone(),
        gift_wrap.pubkey.clone(), // ephemeral pubkey
        gift_wrap.content,
    )?;

    // Parse seal
    let seal = deserialize_event(&seal_json)?;

    // Verify seal kind
    if seal.kind != KIND_SEAL {
        return Err(CryptoError::InvalidCiphertext);
    }

    // Verify seal signature
    let seal_verified = verify_event(seal.clone());

    // Extract sender pubkey from seal
    let sender_pubkey = seal.pubkey.clone();

    // Decrypt seal to get rumor
    let rumor_json = nip44_decrypt(
        recipient_private_key,
        sender_pubkey.clone(),
        seal.content,
    )?;

    // Parse rumor
    let rumor = deserialize_event(&rumor_json)?;

    // Verify rumor kind
    if rumor.kind != KIND_RUMOR {
        return Err(CryptoError::InvalidCiphertext);
    }

    Ok(UnwrapResult {
        rumor,
        sender_pubkey,
        seal_verified,
    })
}

/// Serialize a NostrEvent to JSON
fn serialize_event(event: &NostrEvent) -> Result<String, CryptoError> {
    serde_json::to_string(&EventJsonOut {
        id: &event.id,
        pubkey: &event.pubkey,
        created_at: event.created_at,
        kind: event.kind,
        tags: &event.tags,
        content: &event.content,
        sig: &event.sig,
    })
    .map_err(|_| CryptoError::InvalidJson)
}

/// Deserialize JSON to NostrEvent
fn deserialize_event(json: &str) -> Result<NostrEvent, CryptoError> {
    // Safe parsing - check for prototype pollution
    if json.contains("__proto__") || json.contains("constructor") || json.contains("prototype") {
        return Err(CryptoError::InvalidJson);
    }

    let parsed: EventJsonIn = serde_json::from_str(json).map_err(|_| CryptoError::InvalidJson)?;

    Ok(NostrEvent {
        id: parsed.id,
        pubkey: parsed.pubkey,
        created_at: parsed.created_at,
        kind: parsed.kind,
        tags: parsed.tags,
        content: parsed.content,
        sig: parsed.sig,
    })
}

/// JSON representation for serialization
#[derive(serde::Serialize)]
struct EventJsonOut<'a> {
    id: &'a str,
    pubkey: &'a str,
    created_at: i64,
    kind: i32,
    tags: &'a Vec<Vec<String>>,
    content: &'a str,
    sig: &'a str,
}

/// JSON representation for deserialization
#[derive(serde::Deserialize)]
struct EventJsonIn {
    id: String,
    pubkey: String,
    created_at: i64,
    kind: i32,
    tags: Vec<Vec<String>>,
    content: String,
    sig: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_full_gift_wrap_flow() {
        let sender = generate_keypair();
        let recipient = generate_keypair();

        let message = "Hello, this is a private message!";
        let now = 1700000000i64;

        // Step 1: Create rumor
        let rumor = create_rumor(
            sender.public_key.clone(),
            recipient.public_key.clone(),
            message.to_string(),
            now,
        )
        .unwrap();

        assert_eq!(rumor.kind, KIND_RUMOR);
        assert!(rumor.sig.is_empty()); // Rumor is unsigned

        // Step 2: Create seal
        let seal = create_seal(
            sender.private_key.clone(),
            recipient.public_key.clone(),
            rumor.clone(),
            now,
        )
        .unwrap();

        assert_eq!(seal.kind, KIND_SEAL);
        assert!(!seal.sig.is_empty()); // Seal is signed

        // Step 3: Create gift wrap
        let gift_wrap = create_gift_wrap(
            recipient.public_key.clone(),
            seal.clone(),
            now,
        )
        .unwrap();

        assert_eq!(gift_wrap.kind, KIND_GIFT_WRAP);
        assert!(!gift_wrap.sig.is_empty()); // Gift wrap is signed

        // Verify gift wrap has p tag
        assert_eq!(gift_wrap.tags.len(), 1);
        assert_eq!(gift_wrap.tags[0][0], "p");
        assert_eq!(gift_wrap.tags[0][1], recipient.public_key);

        // Step 4: Unwrap on recipient side
        let result = unwrap_gift_wrap(recipient.private_key.clone(), gift_wrap).unwrap();

        assert!(result.seal_verified);
        assert_eq!(result.sender_pubkey, sender.public_key);
        assert_eq!(result.rumor.content, message);
    }

    #[test]
    fn test_wrong_recipient_fails() {
        let sender = generate_keypair();
        let recipient = generate_keypair();
        let wrong = generate_keypair();

        let message = "Secret message";
        let now = 1700000000i64;

        let rumor = create_rumor(
            sender.public_key.clone(),
            recipient.public_key.clone(),
            message.to_string(),
            now,
        )
        .unwrap();

        let seal = create_seal(
            sender.private_key.clone(),
            recipient.public_key.clone(),
            rumor,
            now,
        )
        .unwrap();

        let gift_wrap = create_gift_wrap(recipient.public_key.clone(), seal, now).unwrap();

        // Try to unwrap with wrong key
        let result = unwrap_gift_wrap(wrong.private_key.clone(), gift_wrap);
        assert!(result.is_err());
    }

    #[test]
    fn test_timestamp_randomization() {
        let sender = generate_keypair();
        let recipient = generate_keypair();

        let now = 1700000000i64;
        let mut timestamps = Vec::new();

        for _ in 0..10 {
            let rumor = create_rumor(
                sender.public_key.clone(),
                recipient.public_key.clone(),
                "Test".to_string(),
                now,
            )
            .unwrap();

            timestamps.push(rumor.created_at);
        }

        // All timestamps should be within ±2 days
        for ts in &timestamps {
            assert!(*ts >= now - TIMESTAMP_RANGE as i64);
            assert!(*ts <= now + TIMESTAMP_RANGE as i64);
        }

        // Not all timestamps should be the same (randomization working)
        let unique: std::collections::HashSet<_> = timestamps.iter().collect();
        assert!(unique.len() > 1);
    }

    #[test]
    fn test_prototype_pollution_prevention() {
        let malicious = r#"{"__proto__": {"isAdmin": true}, "id": "", "pubkey": "", "created_at": 0, "kind": 0, "tags": [], "content": "", "sig": ""}"#;

        let result = deserialize_event(malicious);
        assert!(result.is_err());
    }
}

File: clients/web/src/core/crypto/nip17.ts
import { finalizeEvent, generateSecretKey, verifyEvent } from 'nostr-tools'
import { hexToBytes } from '@noble/hashes/utils'
import { encryptNIP44, decryptNIP44, deriveConversationKey } from './nip44'
import { z } from 'zod'
import type { Rumor, Seal, GiftWrap } from '@/types/nostr'
import { logger } from '@/lib/logger'

/**
 * Normalize a private key to Uint8Array.
 * Accepts either a Uint8Array or a hex-encoded string.
 */
function normalizePrivateKey(key: Uint8Array | string): Uint8Array {
  if (typeof key === 'string') {
    return hexToBytes(key)
  }
  return key
}

/**
 * SECURITY: Zod schemas for validating decrypted NIP-17 content
 *
 * These schemas provide defense-in-depth against:
 * 1. Malformed JSON that passes decryption but causes processing errors
 * 2. Injection attacks via unexpected field types
 * 3. Object prototype pollution via __proto__ or constructor
 */

// Schema for validating Rumor (decrypted inner message)
// Uses passthrough() to allow _v and other protocol fields through validation
const LocalRumorSchema = z.object({
  _v: z.string().default('1.0.0'),
  kind: z.number().int().nonnegative(),
  content: z.string(),
  created_at: z.number().int().nonnegative(),
  tags: z.array(z.array(z.string())),
}).passthrough()

// Schema for validating Seal (signed wrapper)
// Uses passthrough() to allow _v and other protocol fields through validation
const LocalSealSchema = z.object({
  _v: z.string().default('1.0.0'),
  kind: z.literal(13), // Seal kind
  content: z.string(),
  created_at: z.number().int().nonnegative(),
  tags: z.array(z.array(z.string())),
  id: z.string().regex(/^[0-9a-f]{64}$/, 'Invalid event ID format'),
  pubkey: z.string().regex(/^[0-9a-f]{64}$/, 'Invalid pubkey format'),
  sig: z.string().regex(/^[0-9a-f]{128}$/, 'Invalid signature format'),
}).passthrough()

/**
 * SECURITY: Check for prototype pollution attempts in parsed JSON
 *
 * Prototype pollution occurs when attacker-controlled JSON contains
 * __proto__, constructor.prototype, or similar properties that could
 * modify Object.prototype when assigned.
 *
 * We check if these dangerous keys have non-function values, as
 * legitimate JSON shouldn't set these to objects.
 */
function checkPrototypePollution(obj: unknown, path: string = ''): void {
  if (typeof obj !== 'object' || obj === null) {
    return
  }

  // Check if this is an array
  if (Array.isArray(obj)) {
    obj.forEach((item, index) => checkPrototypePollution(item, `${path}[${index}]`))
    return
  }

  // Check for dangerous keys at this level
  const record = obj as Record<string, unknown>

  // __proto__ should never be in JSON
  if (Object.prototype.hasOwnProperty.call(record, '__proto__')) {
    throw new Error(`SECURITY: Prototype pollution attempt detected via __proto__ at ${path}`)
  }

  // prototype property with object value is suspicious
  if (Object.prototype.hasOwnProperty.call(record, 'prototype') && typeof record['prototype'] === 'object') {
    throw new Error(`SECURITY: Prototype pollution attempt detected via prototype at ${path}`)
  }

  // Recursively check nested objects
  for (const [key, value] of Object.entries(record)) {
    if (typeof value === 'object' && value !== null) {
      checkPrototypePollution(value, path ? `${path}.${key}` : key)
    }
  }
}

/**
 * SECURITY: Safe JSON parse with schema validation
 * Prevents prototype pollution and validates structure
 */
function safeJsonParse<T>(json: string, schema: z.ZodSchema<T>, name: string): T {
  let parsed: unknown

  try {
    // First, parse JSON
    parsed = JSON.parse(json)
  } catch (error) {
    throw new Error(`SECURITY: Failed to parse ${name} JSON: ${error instanceof Error ? error.message : 'Invalid JSON'}`)
  }

  // Check for prototype pollution attempts
  checkPrototypePollution(parsed, name)

  // Validate against schema
  const result = schema.safeParse(parsed)
  if (!result.success) {
    const errors = result.error.issues.map(i => `${i.path.join('.')}: ${i.message}`).join('; ')
    throw new Error(`SECURITY: ${name} validation failed: ${errors}`)
  }

  return result.data
}

/**
 * Result of unwrapping a gift wrap, including sender identity
 */
export interface UnwrappedMessage {
  rumor: Rumor;
  senderPubkey: string; // The actual sender (from seal.pubkey)
  sealVerified: boolean; // Whether the seal signature verified
}

/**
 * Generate a cryptographically secure random integer in range [0, max)
 * Uses crypto.getRandomValues() for security-critical randomness
 */
export function secureRandomInt(max: number): number {
  const randomBuffer = new Uint32Array(1)
  crypto.getRandomValues(randomBuffer)
  // Use rejection sampling to avoid modulo bias
  const maxValid = Math.floor(0xFFFFFFFF / max) * max
  if (randomBuffer[0] >= maxValid) {
    // Extremely rare - retry
    return secureRandomInt(max)
  }
  return randomBuffer[0] % max
}

/**
 * Randomize timestamp within a 2-day window for metadata protection
 * As per NIP-17 spec: random timestamp up to 2 days in the PAST only.
 * Never produces future timestamps, which relays reject as "created_at too late".
 * Uses crypto.getRandomValues() to prevent timing correlation attacks.
 */
export function randomizeTimestamp(baseTime: number = Date.now()): number {
  const twoDaysInSeconds = 2 * 24 * 60 * 60
  const randomOffset = secureRandomInt(twoDaysInSeconds)
  return Math.floor(baseTime / 1000) - randomOffset
}

/**
 * Create an unsigned rumor event (step 1 of NIP-17)
 */
export function createRumor(
  kind: number,
  content: string,
  recipientPubkey: string,
  tags: string[][] = []
): Rumor {
  return {
    _v: '1.0.0',
    kind,
    content,
    created_at: randomizeTimestamp(),
    tags: [['p', recipientPubkey], ...tags],
  }
}

/**
 * Create a seal event (step 2 of NIP-17)
 * Seals the rumor with sender's key
 */
export function createSeal(
  rumor: Rumor,
  senderPrivateKey: Uint8Array | string
): Seal {
  const privKeyBytes = normalizePrivateKey(senderPrivateKey)

  // Encrypt the rumor
  const conversationKey = deriveConversationKey(privKeyBytes, rumor.tags.find(t => t[0] === 'p')?.[1] || '')
  const encryptedContent = encryptNIP44(JSON.stringify(rumor), conversationKey)

  // Create and sign the seal
  const sealTemplate = {
    kind: 13,
    content: encryptedContent,
    created_at: randomizeTimestamp(),
    tags: [],
  }

  return finalizeEvent(sealTemplate, privKeyBytes) as unknown as Seal
}

/**
 * Create a gift wrap event (step 3 of NIP-17)
 * Wraps the seal for the recipient using an ephemeral key
 */
export function createGiftWrap(
  seal: Seal,
  recipientPubkey: string
): GiftWrap {
  // Generate ephemeral key for anonymity
  const ephemeralPrivateKey = generateSecretKey()

  // Encrypt the seal for the recipient
  const conversationKey = deriveConversationKey(ephemeralPrivateKey, recipientPubkey)
  const encryptedContent = encryptNIP44(JSON.stringify(seal), conversationKey)

  // Create and sign the gift wrap
  const giftWrapTemplate = {
    kind: 1059,
    content: encryptedContent,
    created_at: randomizeTimestamp(),
    tags: [['p', recipientPubkey]],
  }

  return finalizeEvent(giftWrapTemplate, ephemeralPrivateKey) as unknown as GiftWrap
}

/**
 * Create a complete NIP-17 private DM
 * Returns a gift-wrapped event ready to publish
 */
export function createPrivateDM(
  content: string,
  senderPrivateKey: Uint8Array | string,
  recipientPubkey: string,
  tags: string[][] = []
): GiftWrap {
  // Step 1: Create rumor
  const rumor = createRumor(14, content, recipientPubkey, tags)

  // Step 2: Create seal
  const seal = createSeal(rumor, senderPrivateKey)

  // Step 3: Create gift wrap
  const giftWrap = createGiftWrap(seal, recipientPubkey)

  return giftWrap
}

/**
 * Unwrap and decrypt a NIP-17 gift wrap
 *
 * SECURITY: This function now returns the full UnwrappedMessage including:
 * - The actual sender's pubkey (from seal.pubkey, NOT giftWrap.pubkey)
 * - Verification status of the seal signature
 *
 * The giftWrap.pubkey is an ephemeral key and MUST NOT be trusted as sender identity.
 *
 * SECURITY: Uses schema validation on decrypted content to prevent:
 * - Malformed JSON injection
 * - Prototype pollution attacks
 * - Type confusion attacks
 */
export function unwrapGiftWrap(
  giftWrap: GiftWrap,
  recipientPrivateKey: Uint8Array | string
): UnwrappedMessage {
  const privKeyBytes = normalizePrivateKey(recipientPrivateKey)

  // Decrypt the seal using the ephemeral key in the gift wrap
  const conversationKey = deriveConversationKey(privKeyBytes, giftWrap.pubkey)
  const sealJson = decryptNIP44(giftWrap.content, conversationKey)
  // SECURITY: Validate decrypted seal against schema
  const seal = safeJsonParse<Seal>(sealJson, LocalSealSchema, 'Seal')

  // SECURITY: Verify the seal's signature to ensure sender authenticity
  // The seal.pubkey is the ACTUAL sender, and we must verify they signed it
  let sealVerified = false
  try {
    sealVerified = verifyEvent(seal)
  } catch (error) {
    logger.warn('Seal signature verification failed:', error)
    sealVerified = false
  }

  // Decrypt the rumor from the seal
  const rumorConversationKey = deriveConversationKey(privKeyBytes, seal.pubkey)
  const rumorJson = decryptNIP44(seal.content, rumorConversationKey)
  // SECURITY: Validate decrypted rumor against schema
  const rumor = safeJsonParse<Rumor>(rumorJson, LocalRumorSchema, 'Rumor')

  return {
    rumor,
    senderPubkey: seal.pubkey, // The ACTUAL sender identity
    sealVerified,
  }
}

/**
 * Legacy function for backward compatibility
 * @deprecated Use unwrapGiftWrap which returns full UnwrappedMessage
 */
export function unwrapGiftWrapLegacy(
  giftWrap: GiftWrap,
  recipientPrivateKey: Uint8Array | string
): Rumor {
  const result = unwrapGiftWrap(giftWrap, recipientPrivateKey)
  return result.rumor
}

/**
 * Create multiple gift wraps for group messaging
 * Each member gets their own encrypted copy
 */
export function createGroupMessage(
  content: string,
  senderPrivateKey: Uint8Array | string,
  recipientPubkeys: string[],
  tags: string[][] = []
): GiftWrap[] {
  return recipientPubkeys.map(recipientPubkey =>
    createPrivateDM(content, senderPrivateKey, recipientPubkey, tags)
  )
}

/**
 * Verify if a gift wrap is intended for a specific recipient
 */
export function isGiftWrapForRecipient(
  giftWrap: GiftWrap,
  recipientPubkey: string
): boolean {
  return giftWrap.tags.some(tag => tag[0] === 'p' && tag[1] === recipientPubkey)
}
